cpp/harvest/marc_issn_lookup.cc

```diff
@@ -105,80 +114,184 @@ void UpdateSubfield(MARC::Subfields &subfields, const SubFieldInfo &sub_field_in
     }
 }
 
-
-void UpdateJournalValidity(std::map<std::string, SubFieldInfo> &journal_cache) {
-    // check whether the issn is occur more than once
-    unsigned online_valid(0), online_invalid(0), print_valid(0), print_invalid(0);
-
-    LOG_INFO("Cache contents:");
-    for (auto sfi : journal_cache) {
-        if (sfi.second.online_version_counter_ == 1) {
-            journal_cache[sfi.first].is_valid_ = true;
-            LOG_INFO(sfi.first + ": Online (valid)");
-            ++online_valid;
-        } else if (sfi.second.online_version_counter_ > 1) {
-            journal_cache[sfi.first].is_valid_ = false;
-            LOG_INFO(sfi.first + ": Online (invalid)");
-            ++online_invalid;
-        } else if (sfi.second.online_version_counter_ == 0)
-            if (sfi.second.printed_version_counter_ == 1) {
-                journal_cache[sfi.first].is_valid_ = true;
-                LOG_INFO(sfi.first + ": Print (valid)");
-                ++print_valid;
-            } else {
-                journal_cache[sfi.first].is_valid_ = false;
-                LOG_INFO(sfi.first + ": Print (invalid)");
-                ++print_invalid;
-            }
+void UpdateSubFieldInfo(SubFieldInfo &sfi, const SubFieldInfo &new_sfi, const bool is_online) {
+    sfi.t_ = new_sfi.t_;
+    sfi.w_ = new_sfi.w_;
+    sfi.x_ = new_sfi.x_;
+    sfi.is_online_ = is_online;
+    (is_online ? ++sfi.online_version_counter_ : ++sfi.printed_version_counter_);
+    if (not new_sfi.l_.empty()) {
+        sfi.l_ = new_sfi.l_;
+        sfi.list_of_l_.emplace_back(new_sfi.l_);
     }
-    LOG_INFO("Online (valid): " + std::to_string(online_valid) + ", " + "Online (invalid): " + std::to_string(online_invalid) + ", "
-             + "Print (valid): " + std::to_string(print_valid) + ", " + "Print (invalid): " + std::to_string(print_invalid));
 }
 
-
-std::map<std::string, SubFieldInfo> BuildJournalCache(const std::string &input_journal_filename) {
-    std::map<std::string, SubFieldInfo> journal_cache;
+std::vector<SubFieldInfo> BuildJournalCache(const std::string &input_journal_filename) {
+    std::vector<SubFieldInfo> journal_cache;
     auto input_journal_file(MARC::Reader::Factory(input_journal_filename));
-    int record_counter(0);
+
 
     while (MARC::Record record = input_journal_file->read()) {
         SubFieldInfo sub_field_info_of_record(record);
+        bool is_exist_in_journal_cache(false);
```

Comment:
`is_exist_in_journal_cache` should be renamed to `exists_in_journal_cache`.

To find an existing element, i think it would be easier to just take 022a as well as 022l directly out of the record (without generating sub_field_info_of_record), and check if one of them is already contained in `list_of_l_` (= `issns_`).